File: Design-Patterns-Using-CSharp/ArchitecturalPatterns/PipeAndFilter.cs
using System;
using System.Collections.Generic;

// Filter Base Class
abstract class Filter
{
    public abstract string Process(string data);
}

// Filters
class CapitalizeFilter : Filter
{
    public override string Process(string data)
    {
        return data.ToUpper();
    }
}

class ReplaceSpaceFilter : Filter
{
    public override string Process(string data)
    {
        return data.Replace(" ", "_");
    }
}

class RemoveSpecialCharactersFilter : Filter
{
    public override string Process(string data)
    {
        string specialCharacters = ",@!";
        var result = new System.Text.StringBuilder();
        foreach (char c in data)
        {
            if (specialCharacters.IndexOf(c) == -1)
            {
                result.Append(c);
            }
        }
        return result.ToString();
    }
}

// Data Processing Pipeline
class DataProcessingPipeline
{
    private List<Filter> filters;

    public DataProcessingPipeline()
    {
        this.filters = new List<Filter>();
    }

    public void AddFilter(Filter filter)
    {
        this.filters.Add(filter);
    }

    public string ProcessData(string data)
    {
        foreach (Filter filter in this.filters)
        {
            data = filter.Process(data);
        }
        return data;
    }
}

// Main class
class PipeAndFilter
{
    public static void Main(string[] args)
    {
        DataProcessingPipeline pipeline = new DataProcessingPipeline();
        pipeline.AddFilter(new CapitalizeFilter());
        pipeline.AddFilter(new ReplaceSpaceFilter());
        pipeline.AddFilter(new RemoveSpecialCharactersFilter());

        string data = "Hello, World!";
        string result = pipeline.ProcessData(data);
        Console.WriteLine("Result: " + result);
    }
}
/*
Result: HELLO_WORLD
*/

File: Design-Patterns-Using-Go/ArchitecturalPatterns/PipeAndFilter.go
package main

import (
	"fmt"
	"strings"
)

// Filter interface
type Filter interface {
	Process(data string) string
}

// CapitalizeFilter struct
type CapitalizeFilter struct{}

// Process method for CapitalizeFilter
func (c *CapitalizeFilter) Process(data string) string {
	return strings.ToUpper(data)
}

// ReplaceSpaceFilter struct
type ReplaceSpaceFilter struct{}

// Process method for ReplaceSpaceFilter
func (r *ReplaceSpaceFilter) Process(data string) string {
	return strings.ReplaceAll(data, " ", "_")
}

// RemoveSpecialCharactersFilter struct
type RemoveSpecialCharactersFilter struct{}

// Process method for RemoveSpecialCharactersFilter
func (r *RemoveSpecialCharactersFilter) Process(data string) string {
	var specialCharacters = ",@!"
	var result strings.Builder
	for _, c := range data {
		if !strings.ContainsRune(specialCharacters, c) {
			result.WriteRune(c)
		}
	}
	return result.String()
}

// DataProcessingPipeline struct
type DataProcessingPipeline struct {
	filters []Filter
}

// NewDataProcessingPipeline constructor
func NewDataProcessingPipeline() *DataProcessingPipeline {
	return &DataProcessingPipeline{
		filters: make([]Filter, 0),
	}
}

// AddFilter method for DataProcessingPipeline
func (d *DataProcessingPipeline) AddFilter(filter Filter) {
	d.filters = append(d.filters, filter)
}

// ProcessData method for DataProcessingPipeline
func (d *DataProcessingPipeline) ProcessData(data string) string {
	for _, filter := range d.filters {
		data = filter.Process(data)
	}
	return data
}

// Client Code
func main() {
	pipeline := NewDataProcessingPipeline()
	pipeline.AddFilter(&CapitalizeFilter{})
	pipeline.AddFilter(&ReplaceSpaceFilter{})
	pipeline.AddFilter(&RemoveSpecialCharactersFilter{})

	data := "Hello, World!"
	result := pipeline.ProcessData(data)
	fmt.Println("Result:", result)
}

/*
Result: HELLO_WORLD
*/